// @plentymarkets/terra-components/app/components/tables/data-table/context-menu/terra-data-table-context-menu.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var terra_data_table_context_menu_service_1 = require("./service/terra-data-table-context-menu.service");
/**
 * @author mkunze
 */
var TerraDataTableContextMenuComponent = /** @class */ (function () {
    function TerraDataTableContextMenuComponent(_contextMenuService) {
        var _this = this;
        this._contextMenuService = _contextMenuService;
        this._contextMenuLinkList = [];
        this._isShown = false;
        this._locationCss = {
            visibility: 'hidden',
            left: 0,
            top: 0,
            right: 0
        };
        this._mouseLocation = {
            left: 0,
            top: 0
        };
        _contextMenuService.show.subscribe(function (e) { return _this.showMenu(e.event, e.obj); });
        _contextMenuService.init.subscribe(function (e) { return _this._contextMenuLinkList = e; });
        this.clickListener = function (event) {
            _this.clickedOutside();
            event.stopPropagation();
        };
    }
    TerraDataTableContextMenuComponent.prototype.clickedOutside = function () {
        this._isShown = false;
        this._locationCss = this.calcMenuPosition();
        document.removeEventListener('click', this.clickListener);
    };
    TerraDataTableContextMenuComponent.prototype.showMenu = function (event, contextMenuLinkList) {
        this._isShown = true;
        this._contextMenuLinkList = contextMenuLinkList;
        this._mouseLocation = {
            left: event.clientX,
            top: event.clientY
        };
        this._locationCss = this.calcMenuPosition();
        event.stopPropagation();
        document.addEventListener('click', this.clickListener);
    };
    TerraDataTableContextMenuComponent.prototype.calcMenuPosition = function () {
        // 70 (navbar) + 46 (tabbar) + 36 (breadcrumbs)
        var offsetTop = 161;
        var offsetLeft;
        var anchor = $('.context-menu#menu');
        var isMenuAtBottom;
        var contextMenuHeight = anchor.height();
        var contextMenuWidth = anchor.width();
        var tableWidth;
        if (this._mouseLocation.top + contextMenuHeight > innerHeight) {
            isMenuAtBottom = true;
        }
        var dataTableElement = anchor.closest('terra-data-table');
        offsetLeft = dataTableElement.offset().left;
        tableWidth = dataTableElement.find('tbody').width();
        if (Math.abs(this._mouseLocation.left - offsetLeft - 2) + contextMenuWidth > tableWidth) {
            offsetLeft = offsetLeft + contextMenuWidth - 6;
        }
        return {
            visibility: this._isShown ? 'visible' : 'hidden',
            left: this._mouseLocation.left - offsetLeft - 2 + 'px',
            top: isMenuAtBottom ? this._mouseLocation.top - offsetTop - contextMenuHeight + 'px' : this._mouseLocation.top - offsetTop - 2 + 'px'
        };
    };
    TerraDataTableContextMenuComponent = __decorate([
        core_1.Component({
            selector: 'context-menu-holder',
            styles: [require('./terra-data-table-context-menu.component.scss')],
            template: require('./terra-data-table-context-menu.component.html')
        }),
        __metadata("design:paramtypes", [terra_data_table_context_menu_service_1.TerraDataTableContextMenuService])
    ], TerraDataTableContextMenuComponent);
    return TerraDataTableContextMenuComponent;
}());
exports.TerraDataTableContextMenuComponent = TerraDataTableContextMenuComponent;
