// @plentymarkets/terra-components/app/components/layouts/overlay/example/terra-overlay.component.example.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var terra_overlay_component_1 = require("../terra-overlay.component");
var terra_alert_component_1 = require("../../../alert/terra-alert.component");
var TerraOverlayComponentExample = /** @class */ (function () {
    function TerraOverlayComponentExample() {
        this._addButtonTooltip = 'Hinzufügen';
        this._cancelButtonTooltip = 'Abbrechen';
        this._exampleText = 'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa.' +
            'Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ' +
            'ultricies nec, pellentesque eu, pretium quis, sem. Nulla consequat massa quis enim. Donec pede justo, ' +
            'fringilla vel, aliquet nec, vulputate eget, arcu. In enim justo, rhoncus ut, imperdiet a, venenatis vitae, ' +
            'justo. Nullam dictum felis eu pede mollis pretium. Integer tincidunt. Cras dapibus. Vivamus elementum semper ' +
            'nisi. Aenean vulputate eleifend tellus. Aenean leo ligula, porttitor eu, consequat vitae, eleifend ac, enim. ' +
            'Aliquam lorem ante, dapibus in, viverra quis, feugiat a, tellus. Phasellus viverra nulla ut metus varius ' +
            'laoreet. Quisque rutrum. Aenean imperdiet. Etiam ultricies nisi vel augue. Curabitur ullamcorper ultricies ' +
            'nisi. Nam eget dui. Etiam rhoncus. Maecenas tempus, tellus eget condimentum rhoncus, sem quam semper libero, ' +
            'sit amet adipiscing sem neque sed ipsum. Nam quam nunc, blandit vel, luctus pulvinar, hendrerit id, lorem. ' +
            'Maecenas nec odio et ante tincidunt tempus. Donec vitae sapien ut libero venenatis faucibus. Nullam quis ' +
            'ante. Etiam sit amet orci eget eros faucibus tincidunt. Duis leo. Sed fringilla mauris sit amet nibh. ' +
            'Donec sodales sagittis magna.';
        this.alert = terra_alert_component_1.TerraAlertComponent.getInstance();
    }
    TerraOverlayComponentExample.prototype.ngOnInit = function () {
        var _this = this;
        this.primaryButtonInterface = {
            icon: 'icon-confirm',
            caption: 'Test',
            tooltipText: this.addButtonTooltip,
            isDisabled: false,
            clickFunction: function () { return _this.primaryClicked(_this.viewChildOverlayWithPrimaryButton); }
        };
        this.secondaryButtonInterface = {
            icon: 'icon-cancel',
            caption: 'Cancel',
            tooltipText: this.cancelButtonTooltip,
            isDisabled: false,
            clickFunction: function () { return _this.secondaryClicked(_this.viewChildOverlayWithSecondaryButton); }
        };
        this.PrimaryButtonInterface = {
            icon: 'icon-add',
            caption: 'Test',
            tooltipText: this.addButtonTooltip,
            isDisabled: false,
            clickFunction: function () { return _this.primaryClicked(_this.viewChildOverlayWithButtons); }
        };
        this.SecondaryButtonInterface = {
            icon: 'icon-cancel',
            caption: 'Cancel',
            tooltipText: this.cancelButtonTooltip,
            isDisabled: false,
            clickFunction: function () { return _this.secondaryClicked(_this.viewChildOverlayWithButtons); }
        };
        this.staticPrimaryButtonInterface = {
            icon: 'icon-add',
            caption: 'Test',
            tooltipText: this.addButtonTooltip,
            isDisabled: false,
            clickFunction: function () { return _this.staticSecondaryButtonInterface.isDisabled = false; }
        };
        this.staticSecondaryButtonInterface = {
            icon: 'icon-cancel',
            caption: 'Cancel',
            tooltipText: this.cancelButtonTooltip,
            isDisabled: true,
            clickFunction: function () { return _this.secondaryClicked(_this.viewChildOverlayStatic); }
        };
    };
    TerraOverlayComponentExample.prototype.openOverlayWithoutButtons = function () {
        this.viewChildOverlayWithoutButtons.showOverlay();
    };
    TerraOverlayComponentExample.prototype.openOverlayWithPrimaryButton = function () {
        this.viewChildOverlayWithPrimaryButton.showOverlay();
    };
    TerraOverlayComponentExample.prototype.openOverlayWithSecondaryButton = function () {
        this.viewChildOverlayWithSecondaryButton.showOverlay();
    };
    TerraOverlayComponentExample.prototype.openOverlayWithButtons = function () {
        this.viewChildOverlayWithButtons.showOverlay();
    };
    TerraOverlayComponentExample.prototype.openOverlayStatic = function () {
        this.viewChildOverlayStatic.showOverlay();
    };
    TerraOverlayComponentExample.prototype.primaryClicked = function (overlay) {
        this.alert.addAlert({
            msg: overlay.inputOverlayTitle + ' clicked',
            type: 'success',
            dismissOnTimeout: 0
        });
    };
    TerraOverlayComponentExample.prototype.secondaryClicked = function (overlay) {
        overlay.hideOverlay();
    };
    TerraOverlayComponentExample.prototype.addAlert = function () {
        this.alert.addAlert({
            msg: 'Alert aus einem Overlay',
            type: 'success',
            dismissOnTimeout: 0
        });
    };
    Object.defineProperty(TerraOverlayComponentExample.prototype, "addButtonTooltip", {
        get: function () {
            return this._addButtonTooltip;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TerraOverlayComponentExample.prototype, "cancelButtonTooltip", {
        get: function () {
            return this._cancelButtonTooltip;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TerraOverlayComponentExample.prototype, "primaryButtonInterface", {
        get: function () {
            return this._primaryButtonInterface;
        },
        set: function (value) {
            this._primaryButtonInterface = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TerraOverlayComponentExample.prototype, "secondaryButtonInterface", {
        get: function () {
            return this._secondaryButtonInterface;
        },
        set: function (value) {
            this._secondaryButtonInterface = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TerraOverlayComponentExample.prototype, "PrimaryButtonInterface", {
        get: function () {
            return this._PrimaryButtonInterface;
        },
        set: function (value) {
            this._PrimaryButtonInterface = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TerraOverlayComponentExample.prototype, "SecondaryButtonInterface", {
        get: function () {
            return this._SecondaryButtonInterface;
        },
        set: function (value) {
            this._SecondaryButtonInterface = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TerraOverlayComponentExample.prototype, "staticPrimaryButtonInterface", {
        get: function () {
            return this._staticPrimaryButtonInterface;
        },
        set: function (value) {
            this._staticPrimaryButtonInterface = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TerraOverlayComponentExample.prototype, "staticSecondaryButtonInterface", {
        get: function () {
            return this._staticSecondaryButtonInterface;
        },
        set: function (value) {
            this._staticSecondaryButtonInterface = value;
        },
        enumerable: true,
        configurable: true
    });
    __decorate([
        core_1.ViewChild('viewChildOverlayWithoutButtons'),
        __metadata("design:type", terra_overlay_component_1.TerraOverlayComponent)
    ], TerraOverlayComponentExample.prototype, "viewChildOverlayWithoutButtons", void 0);
    __decorate([
        core_1.ViewChild('viewChildOverlayWithPrimaryButton'),
        __metadata("design:type", terra_overlay_component_1.TerraOverlayComponent)
    ], TerraOverlayComponentExample.prototype, "viewChildOverlayWithPrimaryButton", void 0);
    __decorate([
        core_1.ViewChild('viewChildOverlayWithSecondaryButton'),
        __metadata("design:type", terra_overlay_component_1.TerraOverlayComponent)
    ], TerraOverlayComponentExample.prototype, "viewChildOverlayWithSecondaryButton", void 0);
    __decorate([
        core_1.ViewChild('viewChildAddProcessesToCategoryOverlay'),
        __metadata("design:type", terra_overlay_component_1.TerraOverlayComponent)
    ], TerraOverlayComponentExample.prototype, "viewChildOverlayWithButtons", void 0);
    __decorate([
        core_1.ViewChild('viewChildOverlayStatic'),
        __metadata("design:type", terra_overlay_component_1.TerraOverlayComponent)
    ], TerraOverlayComponentExample.prototype, "viewChildOverlayStatic", void 0);
    TerraOverlayComponentExample = __decorate([
        core_1.Component({
            selector: 'terra-overlay-example',
            template: require('./terra-overlay.component.example.html')
        })
    ], TerraOverlayComponentExample);
    return TerraOverlayComponentExample;
}());
exports.TerraOverlayComponentExample = TerraOverlayComponentExample;
