// @plentymarkets/terra-components/app/components/layouts/taglist/example/terra-taglist.component.example.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var TerraTaglistComponentExample = /** @class */ (function () {
    function TerraTaglistComponentExample() {
        this.tagList = [
            {
                badge: 'tag1'
            },
            {
                badge: 'tag2'
            },
            {
                badge: 'tag3'
            }
        ];
    }
    TerraTaglistComponentExample = __decorate([
        core_1.Component({
            selector: 'terra-taglist-example',
            styles: [require('./terra-taglist.component.example.scss')],
            template: require('./terra-taglist.component.example.html'),
        })
    ], TerraTaglistComponentExample);
    return TerraTaglistComponentExample;
}());
exports.TerraTaglistComponentExample = TerraTaglistComponentExample;
