// @plentymarkets/terra-components/app/terra-components.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var forms_1 = require("@angular/forms");
var http_1 = require("@angular/http");
var ngx_bootstrap_1 = require("ngx-bootstrap");
var ngx_quill_1 = require("ngx-quill");
var terra_components_component_1 = require("./terra-components.component");
var terra_alert_panel_component_1 = require("./components/alert/terra-alert-panel.component");
var terra_text_input_component_1 = require("./components/forms/input/text-input/terra-text-input.component");
var terra_color_picker_component_1 = require("./components/forms/input/color-picker/terra-color-picker.component");
var terra_number_input_component_1 = require("./components/forms/input/number-input/terra-number-input.component");
var terra_button_component_1 = require("./components/buttons/button/terra-button.component");
var terra_tree_component_1 = require("./components/tree/terra-tree.component");
var terra_checkbox_tree_component_1 = require("./components/tree/checkbox-tree/terra-checkbox-tree.component");
var terra_checkbox_component_1 = require("./components/forms/checkbox/terra-checkbox.component");
var terra_radio_button_component_1 = require("./components/forms/radio-button/terra-radio-button.component");
var terra_select_box_component_1 = require("./components/forms/select-box/terra-select-box.component");
var terra_base_toolbar_component_1 = require("./components/toolbar/base-toolbar/terra-base-toolbar.component");
var terra_indicator_component_1 = require("./components/indicator/terra-indicator.component");
var terra_pager_component_1 = require("./components/pager/terra-pager.component");
var terra_info_box_component_1 = require("./components/layouts/info-box/terra-info-box.component");
var terra_taglist_component_1 = require("./components/layouts/taglist/terra-taglist.component");
var terra_tag_component_1 = require("./components/layouts/tag/terra-tag.component");
var terra_loading_spinner_component_1 = require("./components/loading-spinner/terra-loading-spinner.component");
var terra_overlay_component_1 = require("./components/layouts/overlay/terra-overlay.component");
var terra_data_table_component_1 = require("./components/tables/data-table/terra-data-table.component");
var terra_data_table_context_menu_component_1 = require("./components/tables/data-table/context-menu/terra-data-table-context-menu.component");
var terra_data_table_context_menu_directive_1 = require("./components/tables/data-table/context-menu/directive/terra-data-table-context-menu.directive");
var terra_simple_table_component_1 = require("./components/tables/simple/terra-simple-table.component");
var terra_double_input_component_1 = require("./components/forms/input/double-input/terra-double-input.component");
var terra_split_view_component_1 = require("./components/split-view/terra-split-view.component");
var terra_portlet_component_1 = require("./components/layouts/portlet/terra-portlet.component");
var terra_filter_component_1 = require("./components/filter/terra-filter.component");
var terra_multi_select_box_component_1 = require("./components/forms/multi-select-box/terra-multi-select-box.component");
var terra_dynamic_component_loader_component_1 = require("./components/dynamic-component-loader/terra-dynamic-component-loader.component");
var terra_dynamic_module_loader_component_1 = require("./components/dynamic-module-loader/terra-dynamic-module-loader.component");
var terra_tile_box_component_1 = require("./components/tile/box/terra-tile-box.component");
var terra_tile_box_panel_component_1 = require("./components/tile/panel/terra-tile-box-panel.component");
var terra_suggestion_box_component_1 = require("./components/forms/suggestion-box/terra-suggestion-box.component");
var terra_date_picker_component_1 = require("./components/forms/input/date-picker/terra-date-picker.component");
var terra_text_area_input_component_1 = require("./components/forms/input/text-area-input/terra-text-area-input.component");
var terra_card_component_1 = require("./components/layouts/card/terra-card.component");
var terra_navigator_component_1 = require("./components/navigator/terra-navigator.component");
var terra_toggle_component_1 = require("./components/buttons/toggle/terra-toggle.component");
var terra_syntax_editor_component_1 = require("./components/editors/syntax-editor/terra-syntax-editor.component");
var terra_multi_split_view_component_1 = require("./components/split-view/multi/terra-multi-split-view.component");
var terra_slider_component_1 = require("./components/forms/slider/terra-slider.component");
var terra_file_browser_component_1 = require("./components/file-browser/terra-file-browser.component");
var terra_file_input_component_1 = require("./components/forms/input/file-input/terra-file-input.component");
var terra_no_result_notice_component_1 = require("./components/no-result/terra-no-result-notice.component");
var terra_button_with_options_component_1 = require("./components/buttons/button-with-options/terra-button-with-options.component");
var fixed_header_directive_1 = require("./components/tables/fixed-header/fixed-header.directive");
var terra_note_editor_component_1 = require("./components/editors/note-editor/terra-note-editor.component");
var terra_note_component_1 = require("./components/note/terra-note.component");
var terra_node_tree_component_1 = require("./components/tree/node-tree/terra-node-tree.component");
var terra_node_component_1 = require("./components/tree/node-tree/node/terra-node.component");
var platform_browser_1 = require("@angular/platform-browser");
var animations_1 = require("@angular/platform-browser/animations");
var common_1 = require("@angular/common");
var mydatepicker_1 = require("mydatepicker");
var angular_l10n_1 = require("angular-l10n");
var ng2_ace_editor_1 = require("ng2-ace-editor");
var interact_module_1 = require("./components/interactables/interact.module");
var compiler_1 = require("@angular/compiler");
var terra_navigator_split_view_config_1 = require("./components/navigator/config/terra-navigator-split-view.config");
var terra_frontend_storage_service_1 = require("./components/file-browser/terra-frontend-storage.service");
var terra_file_browser_service_1 = require("./components/file-browser/terra-file-browser.service");
var terra_loading_spinner_service_1 = require("./components/loading-spinner/service/terra-loading-spinner.service");
var terra_data_table_context_menu_service_1 = require("./components/tables/data-table/context-menu/service/terra-data-table-context-menu.service");
var terra_base_service_1 = require("./service/terra-base.service");
var terra_url_params_decorator_service_1 = require("./service/data/terra-url-params-decorator.service");
var terra_alert_component_1 = require("./components/alert/terra-alert.component");
var terra_converter_helper_1 = require("./helpers/terra-converter.helper");
var category_tree_config_1 = require("./components/category-picker/config/category-tree.config");
var terra_json_to_form_field_service_1 = require("./components/forms/dynamic-form/service/terra-json-to-form-field.service");
//### import of terra-component examples
var terra_button_component_example_1 = require("./components/buttons/button/example/terra-button.component.example");
var terra_alert_component_example_1 = require("./components/alert/example/terra-alert.component.example");
var terra_card_component_example_1 = require("./components/layouts/card/example/terra-card.component.example");
var terra_base_toolbar_component_example_1 = require("./components/toolbar/base-toolbar/example/terra-base-toolbar.component.example");
var terra_overlay_component_example_1 = require("./components/layouts/overlay/example/terra-overlay.component.example");
var terra_info_box_component_example_1 = require("./components/layouts/info-box/example/terra-info-box.component.example");
var terra_portlet_component_example_1 = require("./components/layouts/portlet/example/terra-portlet.component.example");
var terra_color_picker_component_example_1 = require("./components/forms/input/color-picker/example/terra-color-picker.component.example");
var terra_date_picker_component_example_1 = require("./components/forms/input/date-picker/example/terra-date-picker.component.example");
var terra_double_input_component_example_1 = require("./components/forms/input/double-input/example/terra-double-input.component.example");
var terra_number_input_component_example_1 = require("./components/forms/input/number-input/example/terra-number-input.component.example");
var terra_text_area_input_component_example_1 = require("./components/forms/input/text-area-input/example/terra-text-area-input.component.example");
var terra_text_input_component_example_1 = require("./components/forms/input/text-input/example/terra-text-input.component.example");
var terra_checkbox_component_example_1 = require("./components/forms/checkbox/example/terra-checkbox.component.example");
var terra_multi_select_box_component_example_1 = require("./components/forms/multi-select-box/example/terra-multi-select-box.component.example");
var terra_radio_button_component_example_1 = require("./components/forms/radio-button/example/terra-radio-button.component.example");
var terra_select_box_component_example_1 = require("./components/forms/select-box/example/terra-select-box.component.example");
var terra_slider_component_example_1 = require("./components/forms/slider/example/terra-slider.component.example");
var terra_suggestion_box_component_example_1 = require("./components/forms/suggestion-box/example/terra-suggestion-box.component.example");
var terra_data_table_component_example_1 = require("./components/tables/data-table/example/terra-data-table.component.example");
var terra_tag_component_example_1 = require("./components/layouts/tag/example/terra-tag.component.example");
var terra_file_input_component_example_1 = require("./components/forms/input/file-input/example/terra-file-input.component.example");
var terra_tree_component_example_1 = require("./components/tree/example/terra-tree.component.example");
var terra_checkbox_tree_component_example_1 = require("./components/tree/checkbox-tree/example/terra-checkbox-tree.component.example");
var terra_toggle_component_example_1 = require("./components/buttons/toggle/example/terra-toggle.component.example");
var terra_syntax_editor_component_example_1 = require("./components/editors/syntax-editor/example/terra-syntax-editor.component.example");
var terra_table_simple_component_example_1 = require("./components/tables/simple/example/terra-table-simple.component.example");
var terra_pager_component_example_1 = require("./components/pager/example/terra-pager.component.example");
var terra_taglist_component_example_1 = require("./components/layouts/taglist/example/terra-taglist.component.example");
var terra_navigator_component_example_1 = require("./components/navigator/example/terra-navigator.component.example");
var terra_no_result_notice_component_example_1 = require("./components/no-result/example/terra-no-result-notice.component.example");
var terra_button_with_options_component_example_1 = require("./components/buttons/button-with-options/example/terra-button-with-options.component.example");
var terra_input_component_example_1 = require("./components/forms/input/example/terra-input.component.example");
var terra_filter_component_example_1 = require("./components/filter/example/terra-filter.component.example");
var terra_data_table_context_menu_component_example_1 = require("./components/tables/data-table/context-menu/example/terra-data-table-context-menu.component.example");
var terra_note_editor_component_example_1 = require("./components/editors/note-editor/example/terra-note-editor.component.example");
var terra_note_component_example_1 = require("./components/note/example/terra-note.component.example");
var terra_node_tree_component_example_1 = require("./components/tree/node-tree/example/terra-node-tree.component.example");
var terra_dynamic_form_component_1 = require("./components/forms/dynamic-form/terra-dynamic-form.component");
var terra_dynamic_switch_component_1 = require("./components/forms/dynamic-form/dynamic-switch/terra-dynamic-switch.component");
var terra_dynamic_form_service_1 = require("./components/forms/dynamic-form/service/terra-dynamic-form.service");
var terra_form_field_control_service_1 = require("./components/forms/dynamic-form/service/terra-form-field-control.service");
var terra_multi_check_box_component_1 = require("./components/forms/multi-check-box/terra-multi-check-box.component");
var terra_multi_check_box_component_example_1 = require("./components/forms/multi-check-box/example/terra-multi-check-box.component.example");
var terra_category_picker_component_1 = require("./components/category-picker/terra-category-picker.component");
var TerraComponentsModule = /** @class */ (function () {
    function TerraComponentsModule() {
    }
    TerraComponentsModule_1 = TerraComponentsModule;
    TerraComponentsModule.forRoot = function () {
        return {
            ngModule: TerraComponentsModule_1,
            providers: [
                terra_loading_spinner_service_1.TerraLoadingSpinnerService,
                terra_data_table_context_menu_service_1.TerraDataTableContextMenuService,
                terra_base_service_1.TerraBaseService,
                terra_navigator_split_view_config_1.TerraNavigatorSplitViewConfig,
                terra_url_params_decorator_service_1.TerraUrlParamsDecorator,
                terra_frontend_storage_service_1.TerraFrontendStorageService,
                terra_alert_component_1.TerraAlertComponent,
                terra_dynamic_form_service_1.TerraDynamicFormService,
                terra_form_field_control_service_1.TerraFormFieldControlService,
                terra_json_to_form_field_service_1.TerraJsonToFormFieldService
            ]
        };
    };
    TerraComponentsModule.forChild = function () {
        return {
            ngModule: TerraComponentsModule_1,
            providers: [
                terra_loading_spinner_service_1.TerraLoadingSpinnerService,
                terra_data_table_context_menu_service_1.TerraDataTableContextMenuService,
                terra_base_service_1.TerraBaseService,
                terra_navigator_split_view_config_1.TerraNavigatorSplitViewConfig,
                terra_url_params_decorator_service_1.TerraUrlParamsDecorator,
                terra_frontend_storage_service_1.TerraFrontendStorageService,
                terra_alert_component_1.TerraAlertComponent,
                category_tree_config_1.CategoryTreeConfig,
                terra_json_to_form_field_service_1.TerraJsonToFormFieldService
            ]
        };
    };
    TerraComponentsModule = TerraComponentsModule_1 = __decorate([
        core_1.NgModule({
            declarations: [
                terra_components_component_1.TerraComponentsComponent,
                terra_alert_panel_component_1.TerraAlertPanelComponent,
                terra_text_input_component_1.TerraTextInputComponent,
                terra_color_picker_component_1.TerraColorPickerComponent,
                terra_number_input_component_1.TerraNumberInputComponent,
                terra_button_component_1.TerraButtonComponent,
                terra_tree_component_1.TerraTreeComponent,
                terra_checkbox_tree_component_1.TerraCheckboxTreeComponent,
                terra_checkbox_component_1.TerraCheckboxComponent,
                terra_radio_button_component_1.TerraRadioButtonComponent,
                terra_select_box_component_1.TerraSelectBoxComponent,
                terra_base_toolbar_component_1.TerraBaseToolbarComponent,
                terra_indicator_component_1.TerraIndicatorComponent,
                terra_pager_component_1.TerraPagerComponent,
                terra_indicator_component_1.TerraIndicatorComponent,
                terra_info_box_component_1.TerraInfoBoxComponent,
                terra_taglist_component_1.TerraTaglistComponent,
                terra_tag_component_1.TerraTagComponent,
                terra_loading_spinner_component_1.TerraLoadingSpinnerComponent,
                terra_overlay_component_1.TerraOverlayComponent,
                terra_data_table_component_1.TerraDataTableComponent,
                terra_data_table_context_menu_component_1.TerraDataTableContextMenuComponent,
                terra_data_table_context_menu_directive_1.TerraDataTableContextMenuDirective,
                terra_simple_table_component_1.TerraSimpleTableComponent,
                terra_double_input_component_1.TerraDoubleInputComponent,
                terra_portlet_component_1.TerraPortletComponent,
                terra_split_view_component_1.TerraSplitViewComponent,
                terra_filter_component_1.TerraFilterComponent,
                terra_multi_select_box_component_1.TerraMultiSelectBoxComponent,
                terra_multi_check_box_component_1.TerraMultiCheckBoxComponent,
                terra_dynamic_component_loader_component_1.TerraDynamicComponentLoaderComponent,
                terra_dynamic_module_loader_component_1.TerraDynamicModuleLoaderComponent,
                terra_tile_box_component_1.TerraTileBoxComponent,
                terra_tile_box_panel_component_1.TerraTileBoxPanelComponent,
                terra_suggestion_box_component_1.TerraSuggestionBoxComponent,
                terra_date_picker_component_1.TerraDatePickerComponent,
                terra_text_area_input_component_1.TerraTextAreaInputComponent,
                terra_card_component_1.TerraCardComponent,
                terra_navigator_component_1.TerraNavigatorComponent,
                terra_toggle_component_1.TerraToggleComponent,
                terra_syntax_editor_component_1.TerraSyntaxEditorComponent,
                terra_multi_split_view_component_1.TerraMultiSplitViewComponent,
                terra_slider_component_1.TerraSliderComponent,
                terra_file_browser_component_1.TerraFileBrowserComponent,
                terra_file_input_component_1.TerraFileInputComponent,
                terra_no_result_notice_component_1.TerraNoResultNoticeComponent,
                terra_button_with_options_component_1.TerraButtonWithOptionsComponent,
                fixed_header_directive_1.FixedHeaderDirective,
                terra_note_editor_component_1.TerraNoteEditorComponent,
                terra_note_component_1.TerraNoteComponent,
                terra_node_tree_component_1.TerraNodeTreeComponent,
                terra_node_component_1.TerraNodeComponent,
                terra_dynamic_form_component_1.TerraDynamicFormComponent,
                terra_dynamic_switch_component_1.TerraDynamicSwitchComponent,
                terra_category_picker_component_1.TerraCategoryPickerComponent,
                //### declarations of terra-components examples
                terra_button_component_example_1.TerraButtonComponentExample,
                terra_alert_component_example_1.TerraAlertComponentExample,
                terra_card_component_example_1.TerraCardComponentExample,
                terra_base_toolbar_component_example_1.TerraBaseToolbarComponentExample,
                terra_overlay_component_example_1.TerraOverlayComponentExample,
                terra_info_box_component_example_1.TerraInfoboxComponentExample,
                terra_portlet_component_example_1.TerraPortletComponentExample,
                terra_color_picker_component_example_1.TerraColorPickerComponentExample,
                terra_date_picker_component_example_1.TerraDatePickerComponentExample,
                terra_double_input_component_example_1.TerraDoubleInputComponentExample,
                terra_number_input_component_example_1.TerraNumberInputComponentExample,
                terra_text_area_input_component_example_1.TerraTextAreaInputComponentExample,
                terra_text_input_component_example_1.TerraTextInputComponentExample,
                terra_checkbox_component_example_1.TerraCheckboxComponentExample,
                terra_multi_select_box_component_example_1.TerraMultiSelectBoxComponentExample,
                terra_multi_check_box_component_example_1.TerraMultiCheckBoxComponentExample,
                terra_radio_button_component_example_1.TerraRadioButtonComponentExample,
                terra_select_box_component_example_1.TerraSelectBoxComponentExample,
                terra_slider_component_example_1.TerraSliderComponentExample,
                terra_suggestion_box_component_example_1.TerraSuggestionBoxComponentExample,
                terra_data_table_component_example_1.TerraDataTableComponentExample,
                terra_tag_component_example_1.TerraTagComponentExample,
                terra_file_input_component_example_1.TerraFileInputComponentExample,
                terra_tree_component_example_1.TerraTreeComponentExample,
                terra_checkbox_tree_component_example_1.TerraCheckboxTreeComponentExample,
                terra_toggle_component_example_1.TerraToggleComponentExample,
                terra_syntax_editor_component_example_1.TerraSyntaxEditorComponentExample,
                terra_table_simple_component_example_1.TerraSimpleTableComponentExample,
                terra_pager_component_example_1.TerraPagerComponentExample,
                terra_taglist_component_example_1.TerraTaglistComponentExample,
                terra_navigator_component_example_1.TerraNavigatorComponentExample,
                terra_no_result_notice_component_example_1.TerraNoResultNoticeComponentExample,
                terra_button_with_options_component_example_1.TerraButtonWithOptionsComponentExample,
                terra_input_component_example_1.TerraInputComponentExample,
                terra_filter_component_example_1.TerraFilterComponentExample,
                terra_data_table_context_menu_component_example_1.TerraDataTableContextMenuComponentExample,
                terra_note_editor_component_example_1.TerraNoteEditorComponentExample,
                terra_note_component_example_1.TerraNoteComponentExample,
                terra_node_tree_component_example_1.TerraNodeTreeComponentExample
            ],
            entryComponents: [
                terra_alert_panel_component_1.TerraAlertPanelComponent,
                terra_text_input_component_1.TerraTextInputComponent,
                terra_color_picker_component_1.TerraColorPickerComponent,
                terra_number_input_component_1.TerraNumberInputComponent,
                terra_button_component_1.TerraButtonComponent,
                terra_tree_component_1.TerraTreeComponent,
                terra_checkbox_tree_component_1.TerraCheckboxTreeComponent,
                terra_checkbox_component_1.TerraCheckboxComponent,
                terra_radio_button_component_1.TerraRadioButtonComponent,
                terra_select_box_component_1.TerraSelectBoxComponent,
                terra_base_toolbar_component_1.TerraBaseToolbarComponent,
                terra_indicator_component_1.TerraIndicatorComponent,
                terra_pager_component_1.TerraPagerComponent,
                terra_indicator_component_1.TerraIndicatorComponent,
                terra_info_box_component_1.TerraInfoBoxComponent,
                terra_taglist_component_1.TerraTaglistComponent,
                terra_tag_component_1.TerraTagComponent,
                terra_loading_spinner_component_1.TerraLoadingSpinnerComponent,
                terra_overlay_component_1.TerraOverlayComponent,
                terra_data_table_component_1.TerraDataTableComponent,
                terra_data_table_context_menu_component_1.TerraDataTableContextMenuComponent,
                terra_simple_table_component_1.TerraSimpleTableComponent,
                terra_double_input_component_1.TerraDoubleInputComponent,
                terra_portlet_component_1.TerraPortletComponent,
                terra_split_view_component_1.TerraSplitViewComponent,
                terra_filter_component_1.TerraFilterComponent,
                terra_multi_select_box_component_1.TerraMultiSelectBoxComponent,
                terra_multi_check_box_component_1.TerraMultiCheckBoxComponent,
                terra_tile_box_component_1.TerraTileBoxComponent,
                terra_tile_box_panel_component_1.TerraTileBoxPanelComponent,
                terra_suggestion_box_component_1.TerraSuggestionBoxComponent,
                terra_date_picker_component_1.TerraDatePickerComponent,
                terra_text_area_input_component_1.TerraTextAreaInputComponent,
                terra_card_component_1.TerraCardComponent,
                terra_navigator_component_1.TerraNavigatorComponent,
                terra_toggle_component_1.TerraToggleComponent,
                terra_syntax_editor_component_1.TerraSyntaxEditorComponent,
                terra_multi_split_view_component_1.TerraMultiSplitViewComponent,
                terra_slider_component_1.TerraSliderComponent,
                terra_file_browser_component_1.TerraFileBrowserComponent,
                terra_file_input_component_1.TerraFileInputComponent,
                terra_button_with_options_component_1.TerraButtonWithOptionsComponent,
                terra_note_editor_component_1.TerraNoteEditorComponent,
                terra_note_component_1.TerraNoteComponent,
                terra_node_tree_component_1.TerraNodeTreeComponent,
                terra_node_component_1.TerraNodeComponent,
                terra_category_picker_component_1.TerraCategoryPickerComponent
            ],
            exports: [
                terra_alert_panel_component_1.TerraAlertPanelComponent,
                terra_text_input_component_1.TerraTextInputComponent,
                terra_color_picker_component_1.TerraColorPickerComponent,
                terra_number_input_component_1.TerraNumberInputComponent,
                terra_button_component_1.TerraButtonComponent,
                terra_tree_component_1.TerraTreeComponent,
                terra_checkbox_tree_component_1.TerraCheckboxTreeComponent,
                terra_checkbox_component_1.TerraCheckboxComponent,
                terra_radio_button_component_1.TerraRadioButtonComponent,
                terra_select_box_component_1.TerraSelectBoxComponent,
                terra_base_toolbar_component_1.TerraBaseToolbarComponent,
                terra_indicator_component_1.TerraIndicatorComponent,
                terra_pager_component_1.TerraPagerComponent,
                terra_indicator_component_1.TerraIndicatorComponent,
                terra_info_box_component_1.TerraInfoBoxComponent,
                terra_taglist_component_1.TerraTaglistComponent,
                terra_tag_component_1.TerraTagComponent,
                terra_loading_spinner_component_1.TerraLoadingSpinnerComponent,
                terra_overlay_component_1.TerraOverlayComponent,
                terra_data_table_component_1.TerraDataTableComponent,
                terra_data_table_context_menu_component_1.TerraDataTableContextMenuComponent,
                terra_data_table_context_menu_directive_1.TerraDataTableContextMenuDirective,
                terra_simple_table_component_1.TerraSimpleTableComponent,
                terra_double_input_component_1.TerraDoubleInputComponent,
                terra_portlet_component_1.TerraPortletComponent,
                terra_split_view_component_1.TerraSplitViewComponent,
                terra_filter_component_1.TerraFilterComponent,
                terra_multi_select_box_component_1.TerraMultiSelectBoxComponent,
                terra_multi_check_box_component_1.TerraMultiCheckBoxComponent,
                terra_dynamic_component_loader_component_1.TerraDynamicComponentLoaderComponent,
                terra_dynamic_module_loader_component_1.TerraDynamicModuleLoaderComponent,
                terra_tile_box_component_1.TerraTileBoxComponent,
                terra_tile_box_panel_component_1.TerraTileBoxPanelComponent,
                terra_suggestion_box_component_1.TerraSuggestionBoxComponent,
                terra_date_picker_component_1.TerraDatePickerComponent,
                terra_text_area_input_component_1.TerraTextAreaInputComponent,
                terra_card_component_1.TerraCardComponent,
                terra_navigator_component_1.TerraNavigatorComponent,
                terra_toggle_component_1.TerraToggleComponent,
                terra_syntax_editor_component_1.TerraSyntaxEditorComponent,
                terra_multi_split_view_component_1.TerraMultiSplitViewComponent,
                terra_slider_component_1.TerraSliderComponent,
                terra_file_browser_component_1.TerraFileBrowserComponent,
                terra_file_input_component_1.TerraFileInputComponent,
                terra_no_result_notice_component_1.TerraNoResultNoticeComponent,
                terra_button_with_options_component_1.TerraButtonWithOptionsComponent,
                fixed_header_directive_1.FixedHeaderDirective,
                terra_note_editor_component_1.TerraNoteEditorComponent,
                terra_note_component_1.TerraNoteComponent,
                terra_node_tree_component_1.TerraNodeTreeComponent,
                terra_dynamic_form_component_1.TerraDynamicFormComponent,
                terra_dynamic_switch_component_1.TerraDynamicSwitchComponent,
                terra_category_picker_component_1.TerraCategoryPickerComponent,
                //### declarations of terra-components examples
                terra_button_component_example_1.TerraButtonComponentExample,
                terra_alert_component_example_1.TerraAlertComponentExample,
                terra_card_component_example_1.TerraCardComponentExample,
                terra_base_toolbar_component_example_1.TerraBaseToolbarComponentExample,
                terra_overlay_component_example_1.TerraOverlayComponentExample,
                terra_info_box_component_example_1.TerraInfoboxComponentExample,
                terra_portlet_component_example_1.TerraPortletComponentExample,
                terra_color_picker_component_example_1.TerraColorPickerComponentExample,
                terra_date_picker_component_example_1.TerraDatePickerComponentExample,
                terra_double_input_component_example_1.TerraDoubleInputComponentExample,
                terra_number_input_component_example_1.TerraNumberInputComponentExample,
                terra_text_area_input_component_example_1.TerraTextAreaInputComponentExample,
                terra_text_input_component_example_1.TerraTextInputComponentExample,
                terra_checkbox_component_example_1.TerraCheckboxComponentExample,
                terra_multi_select_box_component_example_1.TerraMultiSelectBoxComponentExample,
                terra_multi_check_box_component_example_1.TerraMultiCheckBoxComponentExample,
                terra_radio_button_component_example_1.TerraRadioButtonComponentExample,
                terra_select_box_component_example_1.TerraSelectBoxComponentExample,
                terra_slider_component_example_1.TerraSliderComponentExample,
                terra_suggestion_box_component_example_1.TerraSuggestionBoxComponentExample,
                terra_data_table_component_example_1.TerraDataTableComponentExample,
                terra_tag_component_example_1.TerraTagComponentExample,
                terra_file_input_component_example_1.TerraFileInputComponentExample,
                terra_tree_component_example_1.TerraTreeComponentExample,
                terra_checkbox_tree_component_example_1.TerraCheckboxTreeComponentExample,
                terra_toggle_component_example_1.TerraToggleComponentExample,
                terra_syntax_editor_component_example_1.TerraSyntaxEditorComponentExample,
                terra_table_simple_component_example_1.TerraSimpleTableComponentExample,
                terra_pager_component_example_1.TerraPagerComponentExample,
                terra_taglist_component_example_1.TerraTaglistComponentExample,
                terra_navigator_component_example_1.TerraNavigatorComponentExample,
                terra_no_result_notice_component_example_1.TerraNoResultNoticeComponentExample,
                terra_button_with_options_component_example_1.TerraButtonWithOptionsComponentExample,
                terra_input_component_example_1.TerraInputComponentExample,
                terra_filter_component_example_1.TerraFilterComponentExample,
                terra_data_table_context_menu_component_example_1.TerraDataTableContextMenuComponentExample,
                terra_note_editor_component_example_1.TerraNoteEditorComponentExample,
                terra_note_component_example_1.TerraNoteComponentExample,
                terra_dynamic_form_component_1.TerraDynamicFormComponent,
                terra_dynamic_switch_component_1.TerraDynamicSwitchComponent,
                terra_node_tree_component_example_1.TerraNodeTreeComponentExample
            ],
            imports: [
                platform_browser_1.BrowserModule,
                animations_1.BrowserAnimationsModule,
                common_1.CommonModule,
                forms_1.FormsModule,
                forms_1.ReactiveFormsModule,
                ngx_bootstrap_1.ModalModule.forRoot(),
                http_1.HttpModule,
                ngx_bootstrap_1.TooltipModule.forRoot(),
                ngx_bootstrap_1.AlertModule.forRoot(),
                ngx_bootstrap_1.ButtonsModule.forRoot(),
                angular_l10n_1.TranslationModule.forRoot(),
                mydatepicker_1.MyDatePickerModule,
                ng2_ace_editor_1.AceEditorModule,
                interact_module_1.TerraInteractModule,
                ngx_quill_1.QuillModule
            ],
            providers: [
                compiler_1.COMPILER_PROVIDERS,
                terra_navigator_split_view_config_1.TerraNavigatorSplitViewConfig,
                terra_frontend_storage_service_1.TerraFrontendStorageService,
                terra_file_browser_service_1.TerraFileBrowserService,
                terra_converter_helper_1.TerraConverterHelper
            ],
            bootstrap: [
                terra_components_component_1.TerraComponentsComponent
            ]
        })
    ], TerraComponentsModule);
    return TerraComponentsModule;
    var TerraComponentsModule_1;
}());
exports.TerraComponentsModule = TerraComponentsModule;
