// @plentymarkets/terra-components/app/components/forms/multi-check-box/example/terra-multi-check-box.component.example.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var TerraMultiCheckBoxComponentExample = /** @class */ (function () {
    function TerraMultiCheckBoxComponentExample() {
        this.disabled1 = false;
        this.error1 = false;
        this.disabled2 = false;
        this.error2 = true;
        this.disabled3 = true;
        this.error3 = false;
    }
    TerraMultiCheckBoxComponentExample.prototype.ngOnInit = function () {
        this.values = [
            {
                value: '0',
                caption: 'Value 1',
                selected: false
            },
            {
                value: '1',
                caption: 'Value 2',
                selected: false
            },
            {
                value: '2',
                caption: 'Value 3',
                selected: false
            },
            {
                value: '3',
                caption: 'Value 4',
                selected: false
            },
            {
                value: '4',
                caption: 'Value 5',
                selected: false
            }
        ];
    };
    TerraMultiCheckBoxComponentExample = __decorate([
        core_1.Component({
            selector: 'terra-multi-check-box-example',
            styles: [require('./terra-multi-check-box.component.example.scss')],
            template: require('./terra-multi-check-box.component.example.html'),
        }),
        __metadata("design:paramtypes", [])
    ], TerraMultiCheckBoxComponentExample);
    return TerraMultiCheckBoxComponentExample;
}());
exports.TerraMultiCheckBoxComponentExample = TerraMultiCheckBoxComponentExample;
