// @plentymarkets/terra-components/app/components/forms/dynamic-form/handler/terra-dynamic-form-functions.handler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @author mfrank
 */
var TerraDynamicFormFunctionsHandler = /** @class */ (function () {
    function TerraDynamicFormFunctionsHandler(saveCallback, savedCallback, errorCallback, onValueChangedCallback) {
        this.saveCallback = saveCallback;
        this.savedCallback = savedCallback;
        this.errorCallback = errorCallback;
        this.onValueChangedCallback = onValueChangedCallback;
    }
    TerraDynamicFormFunctionsHandler.prototype.update = function (formValues) {
        this._formFieldControlService.updateDefaultValues(formValues);
    };
    Object.defineProperty(TerraDynamicFormFunctionsHandler.prototype, "formFieldControlService", {
        set: function (formFieldControlService) {
            this._formFieldControlService = formFieldControlService;
        },
        enumerable: true,
        configurable: true
    });
    return TerraDynamicFormFunctionsHandler;
}());
exports.TerraDynamicFormFunctionsHandler = TerraDynamicFormFunctionsHandler;
