// @plentymarkets/terra-components/app/components/forms/suggestion-box/example/terra-suggestion-box.component.example.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var TerraSuggestionBoxComponentExample = /** @class */ (function () {
    function TerraSuggestionBoxComponentExample() {
        this._iconList = [];
    }
    TerraSuggestionBoxComponentExample.prototype.ngOnInit = function () {
        this._iconClass = "icon-plugin";
        this._iconList.push({
            value: "icon-plugin",
            caption: "icon-plugin"
        }, {
            value: "icon-delete",
            caption: "icon-delete"
        }, {
            value: "icon-add",
            caption: "icon-add"
        }, {
            value: "icon-box_plus",
            caption: "icon-box_plus"
        }, {
            value: "icon-flag_blue",
            caption: "icon-flag_blue"
        });
    };
    TerraSuggestionBoxComponentExample = __decorate([
        core_1.Component({
            selector: 'terra-suggestion-box-example',
            styles: [require('./terra-suggestion-box.component.example.scss')],
            template: require('./terra-suggestion-box.component.example.html'),
        })
    ], TerraSuggestionBoxComponentExample);
    return TerraSuggestionBoxComponentExample;
}());
exports.TerraSuggestionBoxComponentExample = TerraSuggestionBoxComponentExample;
