// @plentymarkets/terra-components/app/components/file-browser/terra-file-browser.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var terra_frontend_storage_service_1 = require("./terra-frontend-storage.service");
var file_browser_split_config_1 = require("./config/file-browser-split.config");
var util_1 = require("util");
var TerraFileBrowserComponent = /** @class */ (function () {
    function TerraFileBrowserComponent(splitConfig, _frontendStorageService) {
        this.splitConfig = splitConfig;
        this._frontendStorageService = _frontendStorageService;
        this.inputAllowedExtensions = [];
        this.inputAllowFolders = true;
        this.outputSelectedChange = new core_1.EventEmitter();
        this.onSelectedUrlChange = new core_1.EventEmitter();
    }
    Object.defineProperty(TerraFileBrowserComponent.prototype, "inputStorageServices", {
        get: function () {
            if (!util_1.isNullOrUndefined(this._storageServices) && this._storageServices.length > 0) {
                return this._storageServices;
            }
            return [this._frontendStorageService];
        },
        set: function (services) {
            this._storageServices = services;
        },
        enumerable: true,
        configurable: true
    });
    TerraFileBrowserComponent.prototype.ngOnInit = function () {
        this.splitConfig.init(this.inputStorageServices);
    };
    TerraFileBrowserComponent.prototype.selectUrl = function (publicUrl) {
        this.onSelectedUrlChange.emit(publicUrl);
    };
    __decorate([
        core_1.Input(),
        __metadata("design:type", Array)
    ], TerraFileBrowserComponent.prototype, "inputAllowedExtensions", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], TerraFileBrowserComponent.prototype, "inputAllowFolders", void 0);
    __decorate([
        core_1.Output(),
        __metadata("design:type", core_1.EventEmitter)
    ], TerraFileBrowserComponent.prototype, "outputSelectedChange", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Array),
        __metadata("design:paramtypes", [Array])
    ], TerraFileBrowserComponent.prototype, "inputStorageServices", null);
    TerraFileBrowserComponent = __decorate([
        core_1.Component({
            selector: 'terra-file-browser',
            template: require('./terra-file-browser.component.html'),
            providers: [file_browser_split_config_1.FileBrowserSplitConfig],
            styles: [
                require('./terra-file-browser.component.scss'),
                require('./terra-file-browser.component.glob.scss').toString()
            ]
        }),
        __metadata("design:paramtypes", [file_browser_split_config_1.FileBrowserSplitConfig,
            terra_frontend_storage_service_1.TerraFrontendStorageService])
    ], TerraFileBrowserComponent);
    return TerraFileBrowserComponent;
}());
exports.TerraFileBrowserComponent = TerraFileBrowserComponent;
