// @plentymarkets/terra-components/app/components/forms/dynamic-form/terra-dynamic-form.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var util_1 = require("util");
var terra_form_field_control_service_1 = require("./service/terra-form-field-control.service");
var terra_dynamic_form_functions_handler_1 = require("./handler/terra-dynamic-form-functions.handler");
var terra_dynamic_form_service_1 = require("./service/terra-dynamic-form.service");
var TerraHtmlMethods;
(function (TerraHtmlMethods) {
    TerraHtmlMethods["GET"] = "get";
    TerraHtmlMethods["POST"] = "post";
    TerraHtmlMethods["PUT"] = "put";
    TerraHtmlMethods["DELETE"] = "delete";
})(TerraHtmlMethods = exports.TerraHtmlMethods || (exports.TerraHtmlMethods = {}));
/**
 * @author mfrank
 */
var TerraDynamicFormComponent = /** @class */ (function () {
    function TerraDynamicFormComponent(_formFieldControlService, _dynamicService) {
        this._formFieldControlService = _formFieldControlService;
        this._dynamicService = _dynamicService;
        this.inputPortletStyle = 'col-xs-12 col-md-4';
        this.inputRequestParams = {
            route: '',
            htmlMethod: null,
            params: {}
        };
    }
    TerraDynamicFormComponent.prototype.ngOnInit = function () {
        if (util_1.isNullOrUndefined(this.inputFormFields) || util_1.isNullOrUndefined(this.inputFormFunctions)) {
            console.error('inputFormFields and inputFormFunctions must be set.');
        }
        else {
            this._formFieldControlService.createFormGroup(this.inputFormFields);
            this.inputFormFunctions.formFieldControlService = this._formFieldControlService;
        }
    };
    TerraDynamicFormComponent.prototype.ngOnChanges = function (changes) {
        if (changes['inputFormFields']) {
            this._formFieldControlService.createFormGroup(this.inputFormFields);
        }
    };
    TerraDynamicFormComponent.prototype.validate = function () {
        if (this._formFieldControlService.dynamicFormGroup.valid) {
            this.inputFormFunctions.saveCallback(this._formFieldControlService.dynamicFormGroup.value);
        }
        else {
            this.inputFormFunctions.errorCallback(this._formFieldControlService.dynamicFormGroup, this._formFieldControlService.translationMapping);
        }
    };
    TerraDynamicFormComponent.prototype.onResetClick = function () {
        this._formFieldControlService.resetForm();
    };
    __decorate([
        core_1.Input(),
        __metadata("design:type", terra_dynamic_form_functions_handler_1.TerraDynamicFormFunctionsHandler)
    ], TerraDynamicFormComponent.prototype, "inputFormFunctions", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Array)
    ], TerraDynamicFormComponent.prototype, "inputFormFields", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], TerraDynamicFormComponent.prototype, "inputPortletStyle", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object)
    ], TerraDynamicFormComponent.prototype, "inputRequestParams", void 0);
    TerraDynamicFormComponent = __decorate([
        core_1.Component({
            selector: 'terra-dynamic-form',
            template: require('./terra-dynamic-form.component.html'),
            providers: [terra_dynamic_form_service_1.TerraDynamicFormService]
        }),
        __metadata("design:paramtypes", [terra_form_field_control_service_1.TerraFormFieldControlService,
            terra_dynamic_form_service_1.TerraDynamicFormService])
    ], TerraDynamicFormComponent);
    return TerraDynamicFormComponent;
}());
exports.TerraDynamicFormComponent = TerraDynamicFormComponent;
