// @plentymarkets/terra-components/app/components/file-browser/config/file-browser-split.config.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var image_preview_module_1 = require("../image-preview/image-preview.module");
var file_list_module_1 = require("../file-list/file-list.module");
var terra_multi_split_view_config_1 = require("../../split-view/multi/data/terra-multi-split-view.config");
var SPLIT_WIDTH_FULL = 'col-xs-12 col-md-12 col-lg-12';
var SPLIT_WIDTH_CONTENT = 'col-xs-12 col-md-9 col-lg-10';
var SPLIT_WIDTH_SIDEBAR = 'col-xs-12 col-md-3 col-lg-2';
var FileBrowserSplitConfig = /** @class */ (function (_super) {
    __extends(FileBrowserSplitConfig, _super);
    function FileBrowserSplitConfig() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    FileBrowserSplitConfig.prototype.init = function (storageServices) {
        var _this = this;
        this._storageServices = storageServices;
        this._fileListView = {
            module: file_list_module_1.TerraFileListModule.forRoot(),
            defaultWidth: SPLIT_WIDTH_CONTENT,
            focusedWidth: SPLIT_WIDTH_FULL,
            name: 'File List',
            mainComponentName: file_list_module_1.TerraFileListModule.getMainComponent(),
            inputs: [
                {
                    name: 'inputStorageServices',
                    value: this._storageServices
                }
            ]
        };
        this.addView(this._fileListView);
        this._imagePreviewView = {
            module: image_preview_module_1.TerraImagePreviewModule.forRoot(),
            defaultWidth: '',
            focusedWidth: SPLIT_WIDTH_SIDEBAR,
            name: 'Image Preview',
            mainComponentName: image_preview_module_1.TerraImagePreviewModule.getMainComponent(),
            inputs: [
                {
                    name: 'inputStorageObject',
                    value: null
                },
                {
                    name: 'inputStorageService',
                    value: null
                }
            ]
        };
        this.addView(this._imagePreviewView, this._fileListView);
        setTimeout((function () {
            _this.setSelectedView(_this._fileListView);
        }).bind(this));
    };
    FileBrowserSplitConfig.prototype.showImagePreview = function (storageObject, storageService) {
        this._imagePreviewView.inputs = [
            {
                name: 'inputStorageService',
                value: storageService
            },
            {
                name: 'inputStorageObject',
                value: storageObject
            }
        ];
        if (storageService.isImagePreviewEnabled) {
            this.setSelectedView(this._imagePreviewView);
        }
    };
    FileBrowserSplitConfig.prototype.hideImagePreview = function () {
        this.setSelectedView(this._fileListView);
    };
    FileBrowserSplitConfig = __decorate([
        core_1.Injectable()
    ], FileBrowserSplitConfig);
    return FileBrowserSplitConfig;
}(terra_multi_split_view_config_1.TerraMultiSplitViewConfig));
exports.FileBrowserSplitConfig = FileBrowserSplitConfig;
