// @plentymarkets/terra-components/app/components/forms/input/double-input/terra-double-input.component.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var terra_input_component_1 = require("../terra-input.component");
var forms_1 = require("@angular/forms");
var _1 = require("../../../../../");
var nextId = 0;
var TerraDoubleInputComponent = /** @class */ (function (_super) {
    __extends(TerraDoubleInputComponent, _super);
    function TerraDoubleInputComponent() {
        var _this = _super.call(this, _1.TerraRegex.DOUBLE) || this;
        /**
         *
         * @description Set the decimal count. Default is 2. (0.01)
         */
        _this.inputDecimalCount = 2;
        // generate the id of the input instance
        _this._id = "double-input_#" + nextId++;
        return _this;
    }
    TerraDoubleInputComponent_1 = TerraDoubleInputComponent;
    Object.defineProperty(TerraDoubleInputComponent.prototype, "inputValue", {
        /**
         * @deprecated
         * @param {number} v
         */
        set: function (v) {
            console.warn('inputValue is deprecated. It will be removed in one of the upcoming releases. Please use ngModel instead.');
            this.value = v;
        },
        enumerable: true,
        configurable: true
    });
    TerraDoubleInputComponent.prototype.ngOnInit = function () {
        this.regex = _1.TerraRegex.getDouble(this.inputDecimalCount);
        this._step = 1 / (Math.pow(10, this.inputDecimalCount));
    };
    /**
     * Set the focus on the native input element.
     */
    TerraDoubleInputComponent.prototype.focusNativeInput = function () {
        var _this = this;
        setTimeout(function () {
            var input = document.getElementById(_this._id);
            input.focus();
        });
    };
    /**
     * Select the content of the native input element.
     */
    TerraDoubleInputComponent.prototype.selectNativeInput = function () {
        var _this = this;
        setTimeout(function () {
            var input = document.getElementById(_this._id);
            input.select();
        });
    };
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], TerraDoubleInputComponent.prototype, "inputIsPriceInput", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Number)
    ], TerraDoubleInputComponent.prototype, "inputDecimalCount", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Number),
        __metadata("design:paramtypes", [Number])
    ], TerraDoubleInputComponent.prototype, "inputValue", null);
    TerraDoubleInputComponent = TerraDoubleInputComponent_1 = __decorate([
        core_1.Component({
            selector: 'terra-double-input',
            styles: [require('./terra-double-input.component.scss')],
            template: require('./terra-double-input.component.html'),
            providers: [
                {
                    provide: forms_1.NG_VALUE_ACCESSOR,
                    useExisting: core_1.forwardRef(function () { return TerraDoubleInputComponent_1; }),
                    multi: true
                }
            ]
        }),
        __metadata("design:paramtypes", [])
    ], TerraDoubleInputComponent);
    return TerraDoubleInputComponent;
    var TerraDoubleInputComponent_1;
}(terra_input_component_1.TerraInputComponent));
exports.TerraDoubleInputComponent = TerraDoubleInputComponent;
