// @plentymarkets/terra-components/app/components/buttons/toggle/terra-toggle.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var forms_1 = require("@angular/forms");
exports.TOGGLE_CONTROL_VALUE_ACCESSOR = {
    provide: forms_1.NG_VALUE_ACCESSOR,
    useExisting: core_1.forwardRef(function () { return TerraToggleComponent; }),
    multi: true
};
var TerraToggleComponent = /** @class */ (function () {
    function TerraToggleComponent() {
        this.deactivated = new core_1.EventEmitter();
        this.activated = new core_1.EventEmitter();
        this.toggled = new core_1.EventEmitter();
        this._isActive = false;
        this.onTouchedCallback = function () {
        };
        this.onChangeCallback = function () {
        };
        this.inputTooltipPlacement = 'top';
    }
    TerraToggleComponent.prototype.toggle = function () {
        if (!this.inputIsDisabled) {
            this._isActive = !this._isActive;
            this.toggled.emit(this._isActive);
            this.onChangeCallback(this._isActive);
            if (this._isActive) {
                this.activated.emit();
            }
            else {
                this.deactivated.emit();
            }
        }
    };
    //From ControlValueAccessor interface
    TerraToggleComponent.prototype.writeValue = function (value) {
        if (value !== this._isActive) {
            this._isActive = value;
        }
    };
    //From ControlValueAccessor interface
    TerraToggleComponent.prototype.registerOnChange = function (fn) {
        this.onChangeCallback = fn;
    };
    //From ControlValueAccessor interface
    TerraToggleComponent.prototype.registerOnTouched = function (fn) {
        this.onTouchedCallback = fn;
    };
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], TerraToggleComponent.prototype, "inputIsSmall", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], TerraToggleComponent.prototype, "inputIsLarge", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], TerraToggleComponent.prototype, "inputIsDisabled", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], TerraToggleComponent.prototype, "inputIcon", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], TerraToggleComponent.prototype, "inputIsAlignRight", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], TerraToggleComponent.prototype, "inputIsHidden", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], TerraToggleComponent.prototype, "inputTooltipText", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], TerraToggleComponent.prototype, "inputTooltipPlacement", void 0);
    __decorate([
        core_1.Output(),
        __metadata("design:type", Object)
    ], TerraToggleComponent.prototype, "deactivated", void 0);
    __decorate([
        core_1.Output(),
        __metadata("design:type", Object)
    ], TerraToggleComponent.prototype, "activated", void 0);
    __decorate([
        core_1.Output(),
        __metadata("design:type", Object)
    ], TerraToggleComponent.prototype, "toggled", void 0);
    TerraToggleComponent = __decorate([
        core_1.Component({
            selector: 'terra-toggle',
            styles: [require('./terra-toggle.component.scss')],
            template: require('./terra-toggle.component.html'),
            providers: [exports.TOGGLE_CONTROL_VALUE_ACCESSOR]
        }),
        __metadata("design:paramtypes", [])
    ], TerraToggleComponent);
    return TerraToggleComponent;
}());
exports.TerraToggleComponent = TerraToggleComponent;
